cedar-lean/Cedar/Thm/SymCC/Compiler/WellTyped.lean

```diff
@@ -0,0 +1,2231 @@
+  t.WellFormed εnv.entities
+:= by
+  intros h hcomp; rcases h with ⟨_, _, _, hwf⟩
+  have htwf := compile_wf hwf hcomp
+  simp [htwf]
+
+/--
+Special case for literals
+-/
+theorem compile_well_typed_lit {p : Prim} {tx : CedarType} {Γ : Environment} {εnv : SymEnv} :
+  CompileWellTypedCondition (.lit p tx) Γ εnv →
+  CompileWellTypedForExpr (.lit p tx) εnv
+:= by
+  intros h; rcases h with ⟨_, hεnv, hwt, hwf⟩
+  simp [TypedExpr.typeOf, TypedExpr.toExpr, CompileWellTypedForExpr] at *
+
+  unfold SymEnv.WellFormedFor at hwf
+  rcases hwf with ⟨_, ⟨hrefs⟩⟩
+  rcases hwt with ⟨_, ⟨⟨hwt⟩⟩⟩
+
+  all_goals
+    simp [compile, compilePrim, Factory.someOf]
+    simp [Prim.ValidRef] at hrefs
+    try simp [hrefs]
+    simp [
+      TermType.ofType,
+      Term.typeOf,
+      TermPrim.typeOf,
+      BitVec.width,
+    ]
+
+/- Lemmas that TermType.ofType produces the same result w/ or w/o liftBoolTypes. -/
+mutual
+  private theorem ofQualifiedType_ignores_liftBool
+    {qty : QualifiedType} :
+    TermType.ofQualifiedType qty =
+    TermType.ofQualifiedType qty.liftBoolTypes
+  := by
+    cases qty
+    all_goals
+      simp [
+        TermType.ofQualifiedType,
+        CedarType.liftBoolTypes,
+        QualifiedType.liftBoolTypes,
+      ]
+      apply ofType_ignores_liftBool
+
+  private theorem ofRecordType_ignores_liftBool:
+    ∀ recs : List (Attr × QualifiedType),
+    TermType.ofRecordType recs =
+    TermType.ofRecordType (CedarType.liftBoolTypesRecord recs)
+  | [] => by simp [TermType.ofRecordType, CedarType.liftBoolTypesRecord]
+  | _ :: tail => by
+    simp [TermType.ofRecordType, CedarType.liftBoolTypesRecord]
+    constructor
+    apply ofQualifiedType_ignores_liftBool
+    apply ofRecordType_ignores_liftBool tail
+
+  private theorem ofType_ignores_liftBool :
+    ∀ ty : CedarType,
+    TermType.ofType ty = TermType.ofType ty.liftBoolTypes
+  | .bool _ => by simp [TermType.ofType, CedarType.liftBoolTypes]
+  | .int => by simp [TermType.ofType, CedarType.liftBoolTypes]
+  | .string => by simp [TermType.ofType, CedarType.liftBoolTypes]
+  | .entity ety => by simp [TermType.ofType, CedarType.liftBoolTypes]
+  | .ext xty => by simp [TermType.ofType, CedarType.liftBoolTypes]
+  | .set ty => by
+    simp [TermType.ofType, CedarType.liftBoolTypes]
+    apply ofType_ignores_liftBool ty
+  | .record rty => by
+    simp [TermType.ofType, CedarType.liftBoolTypes, RecordType.liftBoolTypes]
+    apply ofRecordType_ignores_liftBool
+end
+
+theorem isCedarRecordType_implies_isRecordType
+  {ty : TermType} :
+  TermType.isCedarRecordType ty →
+  TermType.isRecordType ty
+:= by
+  intros hty
+  simp [
+    TermType.isCedarRecordType,
+    TermType.isRecordType,
+    TermType.cedarType?,
+  ] at *
+  cases ty
+
+  split at hty
+  split
+
+  any_goals simp
+  any_goals contradiction
+  all_goals simp [TermType.cedarType?] at *
+
+  case prim _ h _ _ =>
+    unfold TermType.cedarType? at *
+    split at h
+    any_goals simp at h
+    any_goals contradiction
```

Comment:
here and in some other spots, I think it would be preferable to use `·` or explicit case names rather than `any_goals`, particularly if the tactic you're using only applies to one goal anyway.  If something changes in our definitions, it will be easier to figure out what went wrong if the proof contains an explicit record of which tactic was "supposed" to apply to which goal.